src/proxy/server.rs

```diff
@@ -27,44 +27,40 @@ use super::tcp;
 /// This type can `serve` new connections, determine what protocol
 /// the connection is speaking, and route it to the corresponding
 /// service.
-pub struct Server<S, B, G>
+pub struct Server<M, B, G>
 where
-    S: NewService<Request=http::Request<HttpBody>>,
-    S::Future: 'static,
+    M: MakeClient<Arc<ServerCtx>, Error = ()> + Clone,
+    M::Client: Service<
+        Request = http::Request<HttpBody>,
+        Response = http::Response<B>,
+    >,
     B: tower_h2::Body,
+    G: GetOriginalDst,
 {
     disable_protocol_detection_ports: IndexSet<u16>,
     drain_signal: drain::Watch,
     get_orig_dst: G,
     h1: hyper::server::conn::Http,
-    h2: tower_h2::Server<
-        HttpBodyNewSvc<S>,
-        ::logging::ServerExecutor,
-        B
-    >,
+    h2_settings: h2::server::Builder,
     listen_addr: SocketAddr,
-    new_service: S,
+    make_client: M,
     proxy_ctx: ProxyCtx,
     transport_registry: transport::metrics::Registry,
     tcp: tcp::Forward,
     log: ::logging::Server,
 }
 
-impl<S, B, G> Server<S, B, G>
+impl<M, B, G> Server<M, B, G>
 where
-    S: NewService<
+    M: MakeClient<Arc<ServerCtx>, Error = ()> + Clone,
+    M::Client: Service<
         Request = http::Request<HttpBody>,
-        Response = http::Response<B>
-    > + Clone + Send + 'static,
-    S::Future: 'static,
-    <S as NewService>::Service: Send,
-    <<S as NewService>::Service as ::tower_service::Service>::Future: Send,
-    S::InitError: fmt::Debug,
-    S::Future: Send + 'static,
-    B: tower_h2::Body + 'static,
-    S::Error: Error + Send + Sync + 'static,
-    S::InitError: Send + fmt::Debug,
-    B: tower_h2::Body + Send + Default + 'static,
+        Response = http::Response<B>,
+    >,
+    M::Client: Send + 'static,
+    <M::Client as Service>::Error: error::Error + Send + Sync + 'static,
+    <M::Client as Service>::Future: Send + 'static,
+    B: tower_h2::Body + Default + Send + 'static,
```

Comment:
Whoa, does this change actually _reduce_ the number of bounds on `Server`? Nice!